[WIP] Gatsby starter

AWolf81 opened the pull request:
**Related Issue:**
#47 

**Summary:**
Add a Gatsby Starter selection to `CreateNewProjectWizard` & add the business logic to pass the starter url to the create command.
I've enlarged the TwoPanelModal height so the new component will fit into it.

There is still a lot todo. But I'd like to check if this is going into the right direction.

The thing with selecting more stuff is something we're handling with Guppy plugins. So we can add Flow, Storybook, Docz,... after project creation to keep the setup small.

**Screenshots/GIFs:**
**Gatsby with starter** (Styling of ProjectStarter input needs to be improved)
![grafik](https://user-images.githubusercontent.com/3046542/48741836-f10b5c00-ec5c-11e8-99ee-d65dc67baf89.png)

**Gatsby starter Summary pane**
![grafik](https://user-images.githubusercontent.com/3046542/48741987-6545ff80-ec5d-11e8-9d67-9398f7e4d913.png)

**Starter selection list (Styling needs to be improved)**
![grafik](https://user-images.githubusercontent.com/3046542/48741805-d0db9d00-ec5c-11e8-8d3a-1317a8612ac1.png)


**Todos:**
- [ ] Improve UI
  - [ ] Create a TextComponent with a button on the right to open the selection menu - similar to `DirectroyPicker`
  - [ ] Use Card component to render Starter selection list & add more details from the yaml.
  - [ ] Improve styling of the selected starter (just a red border at the moment)
  - [ ] Add basic navigation to starter list (just if we're not using Algolia)
- [ ] Refactor new step into `ProjectStarterSelector` component and use it in `projectSpecificSteps` method
- [ ] Selection list
  - [ ] Fix selection updates in `SelectStarterDialog` - Not working at the moment.
  - [ ] Sort by Github stars - requires an additional query. Maybe we need to use Graphql to get all the data. Like in this [code](https://github.com/gatsbyjs/gatsby/blob/f335f0c59e935770404a8844e37900308570ece4/www/src/pages/starters.js).
- [ ] Add test for `config-variables.service`
- [ ] Refactor build steps - I'd like to improve the conditons for button disable & submit button.
- [ ] Add a better copy to the Gatsby Starter summary pane. We'll need more details for the why and what it will do.
- [ ] Show a message if starter couldn't be used. Fails silently at the moment.
- [ ] Check if we need a loading spinner for the selection list.
- [ ] Should we display a preview screenshot? I think there is no image in the `starter.yml` not sure where to get the image - I'll check the Gatsby starter selection page.

**Discussion**
- What features are required in the starter selection dialog? I think we're not needing every feature from [Gatsby starter list](https://www.gatsbyjs.org/starters/?) - maybe we could add a link to it. But the Algolia search would be nice and maybe a sorting by the most popluar starters.
- Should we use Algolia for searching? I think it's OK to use a typeahead component as there are just arount 70 items in the list. We could change to Algolia later.
- Should we also add the `Deploy to Netlify` button? Or is this confusing for beginners? I think it's easy to add but I'm not sure if we should add it.  (Just opening a link like https://app.netlify.com/start/deploy?repository=https://github.com/gatsbyjs/gatsby-starter-blog will deploy to netlify).
- Should we allow the user to enter a URL manually into the starter input? Do we need some prefixes in the input field? For now I'm adding `https://github.com/gatsbyjs/` if no url is entered - so entering `gatsby-starter-blog` will use the starter `https://github.com/gatsbyjs/gatsby-starter-blog`

Haroenv commented on src/components/CreateNewProjectWizard/Gatsby/SelectStarterDialog.js:
```diff
@@ -0,0 +1,233 @@
+
+  componentDidMount() {
+    fetch(
+      'https://raw.githubusercontent.com/gatsbyjs/gatsby/master/docs/starters.yml'
```
to put this in Algolia, we will have to push to Algolia somewhere https://github.com/gatsbyjs/gatsby/blob/master/www/gatsby-node.js here I think, but I'm not too sure yet what the most efficient way will be.

AWolf81 replied:
Thanks, for pointing to the Graphql query in the Gatsby starter. It would be great if we could use a similar query but I'm not sure how to query the same data.
I think for now we can just use the `starter.yml` and I think the selection is OK now. I'll update the screenshots above in a minute.

Also the search can be easily done by filtering through the yml data. So Algolia is not really needed here at the moment.

joshwcomeau commented on src/components/CreateNewProjectWizard/Gatsby/SelectStarterDialog.js:
```diff
@@ -0,0 +1,302 @@
+            starters library{' '}
+            <ExternalLink href="https://www.gatsbyjs.org/starters/">
+              here.
+            </ExternalLink>
```
I think it's slightly more conventional to make the link inline:

```
For a better overview you can also have a look at the{' '}
<ExternalLink href="https://www.gatsbyjs.org/starters/">
  Gatsby starters library
</ExternalLink>.

```

AWolf81 replied:
OK, modified. 
What I've noticed with the external links in the summary pane is that the link is not fully clickable. I think that's also happening on master so this is not introduced here. 
Also the link visiblity is pretty bad as it's not visible that it's clickable. What do you think would it be OK to add a thin light rounded box around these links work?
Or maybe underlining would also work similar to the `import project` button.

For the hover I need to check what's causing it - looks like the width of the container is not wide enough. I haven't found the cause yet.

Here is how this looks:
![screenrecording_externallink_hover](https://user-images.githubusercontent.com/3046542/48999812-077f4f00-f158-11e8-85d8-06b30b0d9073.gif)

AWolf81 replied:
Hover issue fixed with PR #337. It was an issue with the backface pointer events. 
We could improve styling later - thin outline not added.

joshwcomeau commented on src/components/CreateNewProjectWizard/MainPane.js:
```diff
@@ -29,29 +31,116 @@ type Props = {
-class MainPane extends PureComponent<Props> {
+type State = {
+  gatsbyStarter: string, // Temporary value during selection in selection toast
+};
```
hm it feels like this state lives in so many places... Couldn't this invoke `updateFieldValue` instead of being set locally?

AWolf81 replied:
Yes, that's right - it lives in many places. But I'm not sure if I can simplify this. As the user can open the starter selection component and select a starter but then decides to cancel that selection so the previously entered starter is still available.

If I would use `updateFieldValue` that flow wouldn't work.

joshwcomeau commented on src/components/CreateNewProjectWizard/MainPane.js:
```diff
@@ -29,29 +31,116 @@ type Props = {
+    const renderedSteps: Array<?React$Node> = buildSteps.slice(
+      0,
+      currentStepIndex
+    );
```
I feel like this might be a bit over-engineered... I think I'd opt for a "dumber" approach:

```
const steps = [];

if (currentStepIndex > 0) {
  steps.push(/* first step */);
}

if (currentStep > 1 && projectType === 'gatsby') {
  steps.push(/* project-specific step for Gatsby */);
}

if (currentStep > 2) {
  steps.push(/* third step */);
}
```

In addition to being a bit easier to follow, it doesn't create a bunch of elements that might not be used.

AWolf81 replied:
I think you're right. That's a bit over-engineered. :-) Your code is probably easier to follow. I'll have a look at it and maybe I re-write this but I have to think about how I'll implement your proposed flow with-out the additional modal.
So if I'm changing the step order I have to modify this part.

joshwcomeau commented on src/components/CreateNewProjectWizard/MainPane.js:
```diff
@@ -29,29 +31,116 @@ type Props = {
+    return (
+      (currentStepIndex > 0 && !projectType) ||
+      (currentStepIndex >= lastIndex && !projectIcon)
+    );
```
I'd love for this to be broken into a couple steps, so that the steps can be given variable names to convey meaning.

```js
const needsProjectType = !projectType && currentStepIndex > 0;
const needsProjectIcon = !projectIcon && currentStepIndex >= lastIndex;

return needsProjectType || needsProjectIcon;
```

Also, I was initially confused because I assumed `validateField` would return `true` if the field was invalid... how about `isSubmitDisabled` as the name?

AWolf81 replied:
Good idea. Renaming to `isSubmitDisabled` makes it easier to understand. Splitting is also OK as it improves readability. 
I first though that it's clear what these checks are doing but for new devs it's better to have that additional naming.

joshwcomeau commented on src/components/CreateNewProjectWizard/SummaryPane.js:
```diff
@@ -185,6 +185,29 @@ class SummaryPane extends PureComponent<Props> {
       }
+      case 'projectStarter': {
+        // todo: why is a key needed on FadeIn? Was s3t.
+        // todo: should we rename projectStarter to be mores specific as this is Gatbsy only.
```
Typo, Gatsby.

Also I don't understand the first todo, what's s3t?

AWolf81 replied:
Oh, OK, I'll fix that typo.
The `key="s3t"` is in line 178 of SummaryPane. I haven't added it to the Gatsby part. Is the key needed or do I miss why it's there?

joshwcomeau commented on src/components/TextInputWithButton/TextInputWithButton.js:
```diff
@@ -0,0 +1,63 @@
+        >
+          <HoverableOutlineButton
+            noPadding
+            onMouseDown={() => window.requestAnimationFrame(handleFocus)}
```
Why `requestAnimationFrame`?

joshwcomeau commented on src/services/config-variables.service.js:
```diff
@@ -0,0 +1,49 @@
+
+// We're using "template" variables inside the project type configuration file (config/project-types.js)
+// so with the following function we can replace the string $port with the real port number e.g. 3000
+// (see type VariableMap for used mapping strings)
```
Hm, what's the advantage to doing it this way? We used to just pass the variables along to the function instead of needing to do substitution. Did adding starters increase the complexity here?

AWolf81 commented on src/components/CreateNewProjectWizard/Gatsby/ProjectStarterSelection.js:
```diff
@@ -0,0 +1,78 @@
+        gatsbyStarter: selectedStarter,
+      },
+      () => {
+        console.log('updated', this.state);
```
Yes, that's right. Sorry, I've missed that debug stuff. I'll remove it.

AWolf81 commented on src/components/CreateNewProjectWizard/Gatsby/ProjectStarterSelection.js:
```diff
@@ -0,0 +1,78 @@
+
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    return {
+      gatsbyStarter: nextProps.projectStarter,
```
I think you're right but with-out this it would require me to modify modal state as I need to have the info about the entered starter before opening the modal.
So the user can dismiss the changes made in the modal. Anyway I'm not changing this as I'll check if I can remove the second modal - so this won't be needed anymore.

AWolf81 replied:
It's acutally an anti-pattern to override local state unconditionally with every prop change.
There is a nice post [you-probably-dont-need-derived-state](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) about this in the React.js blog.

We don't need it here anymore as I'm rewriting this. But I wanted to note this as I've learned this and I wasn't aware of it.

AWolf81 commented:
I've tried to modify the steps as you mentioned (I'll push it in a minute) but I think there isn't enough space to add the starter behaviour as the last step (see screenshot below). I'd like to keep the notice to the Gatbsy starters, a input for searching (we can use the input we're already having) and a list with the search results + the Codesandbox preview buttons (maybe with a smaller icon).

Maybe we could reduce the spacing between the steps. I'll have a look at it later today.

![grafik](https://user-images.githubusercontent.com/3046542/48974376-f39ff400-f056-11e8-8245-7bc979092e8a.png)

AWolf81 commented:
@joshwcomeau Please have a look at my updated code. I've addressed your review comments.

I've added everything to the same modal & changed the order of the steps.
Spacing between the steps is reduced but I think it's still OK.
Starter section looks OK (see updated screenshot). It would be better to have more than two starters visible in the list but I think there isn't more space available. Fontsize of the starter selection list is pretty small but similar to the other fonts used.
Typeahead and highlighting is also improved and I like how it's working now.

Just the height jumps between the project types but I think that's OK or should we do equal heights? So CRA/Nextjs pane has the same height as Gatsby.

Please let me know if I missed anything from your comments.

AWolf81 commented on src/components/FormField/FormField.js:
```diff
@@ -41,13 +49,18 @@ const getTextColor = (props: Props) => {
-const WrapperLabel = styled.label`
+const WrapperLabel = styled.label.attrs({
+  // default to 30px spacing between FormFields
+  spacing: props => props.spacing || 30,
```
I'll move this to `FormField` component `static defaultProps` as it's only one location to add the default of 30 instead of every styled-component.

AWolf81 commented:
@joshwcomeau could you please have a look if this is working as expected? 
If anything is open, please let me know.

I think just the panel height could be improved but maybe we could change this in a following PR as this is already pretty large.

AWolf81 commented:
@melanieseltzer could you please have a look at this? I think this is ready.
I've change the height of the modal so it's the same for all project types now. I think that's better.

melanieseltzer commented:
This is looking pretty great 👍 I only took a surface level look so far to play around with it, haven't dug into the code yet. Seems like the select starter functionality is working, and the preview in Codesandbox functionality. Since this is quite a large PR I'm going to have to dedicate some more time to take a deeper look, but for now I have some preliminary notes:

- The background is scrolling when you are scrolling inside the starter section and it's a little jarring:

<img src="https://user-images.githubusercontent.com/17421347/51370036-8040be80-1aaa-11e9-8ba0-596b5c74b7d2.gif" width="500">

I wonder if we could leverage the `isVisible` prop in `CreateNewProjectWizard` to disable background scrolling when the modal is open? Maybe with a conditional `createGlobalStyle` once styled-components v4 lands in #269? I'm imagining something like the following, but let me know if you have another idea:

```js
<BackgroundScroll disabled={isVisible}  />

...

const BackgroundScroll = createGlobalStyle`
  html, body {
    overflow: ${props => (props.disabled ? 'hidden' : 'auto')};
  }
`
```


- I'm getting a weird lint error in the dev console from one of the dependencies, do you get this?

```
webpackHotDevClient.js:138 ./node_modules/encoding/lib/iconv-loader.js
9:10-32 Critical dependency: the request of a dependency is an expression
```

- I have some UX concerns with the way a starter is selected:

<img width="386" alt="screen shot 2019-01-17 at 10 35 56 pm" src="https://user-images.githubusercontent.com/17421347/51369396-479fe580-1aa8-11e9-8117-41aba9277052.png">

So once the starter gets selected it gets pulled up to the top of the list, and out of the flow. This is somewhat confusing to me and I feel like it should stay in place and just be highlighted to indicate it's selected. Otherwise the user loses their place and has to scroll back down to get back to where they were, in case they want to make another selection.

Actually I think there's an ordering bug when you make a selection 🤔 Unless I'm misinterpreting this flow, but it doesn't feel intentional. You can see what I mean in this GIF:

<img src="https://user-images.githubusercontent.com/17421347/51369617-1673e500-1aa9-11e9-9e82-aaef871d0cfe.gif" width="500">

As you can see in the screen cap, I scroll to the bottom of the current list of items, and you can see the last two items are `gatsby-starter-mate` and `gatsby-minimal-portfolio-blog`. So when I choose `gatsby-minimal-portfolio-blog` it gets pulled to the top, but when I scroll back down I'm expecting `gatsby-starter-mate` to be the last in the list. But it's not, it's nowhere to be found in that positioning.

It seems like maybe the pagination is off?

AWolf81 commented:
@melanieseltzer Thanks for your review.

- Background scrolling with open modal: Maybe we could disable `pointer-events` on app wrapper if the modal is active to avoid scrolling. I'll have a look at this later today.
- Linting message: No I haven't seen that. I'll check if I'm getting that message too.
- UX issue: The problem here is that we're reusing the input field because of the limited space. Before you click the text entered will be used to filter the list "best match" first item. Once you selected a starter that's the best match and the other result are having a similar url like the picked starter. 
  
  Maybe we can use a different property to filter the typeahead list or do you have an idea how we could improve this? I like to show the picked starter url in the input field but maybe we could just leave the entered typeahead string in the field and mark the selected with the border - but it could be a problem as the selected starter could disappear if the typeahead is changed. 
  (Probably forcing the selected item to be in the list would be a work-around e.g. manually at as first item if not in list)

melanieseltzer commented:
> Maybe we could disable pointer-events on app wrapper if the modal is active to avoid scrolling.

👍

> UX issue: Once you selected a starter that's the best match and the other result are having a similar url like the picked starter.

Ah, gotcha. That's the thing I wasn't understanding about the functionality. As someone just jumping into the flow, I didn't associate selecting a starter (clicking on its name) as setting a best match... I just thought I was making a selection choice, and not expecting the results to change based on that. That's why it's appearing jumbled to me.

Is there a reason to have a 'best match' functionality? As a user I'm expecting search results to be pretty static:

- If I type the word 'blog', all blog related items appear (and I can paginate through them sequentially)
- If I click on one to make it the selected item, I'm not expecting the order of the items to change, or for different results to appear. I'm just making a selection.
- If I change my mind and want to choose a starter before or after the current one,  I should be able to go back to the position in the original list and do that. But with 'best match' the items may have completely re-ordered themselves, or an item previously there isn't even visible anymore, if it's not related to the current 'best match'. E.g. say I click on a starter with 'TypeScript' but I previously saw a starter with 'Redux' that I think would work better. I scroll down to try find the Redux starter again but it's gone because the results are now best match for 'TypeScript' keyword. Sure I could type Redux in search but feels a little clunky to do this...

> I like to show the picked starter url in the input field

It is somewhat problematic for the input text to change to the URL when a selection is made. Somewhat annoying to have typed 'blog' and then after the selection is made, the text is wiped. Cause if you make a selection prematurely but change your mind, you'll have to type 'blog' in again and it's an extra step.

**tl;dr**

Keeping in mind I'm not a UX expert or anything, just going by how I interact with things. So @joshwcomeau and others please feel free to jump in here 🙂 

I think I would suggest:

- No 'best match' functionality. Clicking on a starter simply selects it in place (purple outline) but doesn't move it to the top of the list, and doesn't cause a re-order of results.
- Keep search text in the input, don't put the selected url in there.
- If there needs to be more highlight calling attention to the selected starter, perhaps a tick icon or something? Bad photoshop...

<img src="https://user-images.githubusercontent.com/17421347/51408887-02fe6380-1b15-11e9-80f6-2c92acb71076.jpg" width="400">

AWolf81 commented:
@melanieseltzer you're right, the re-ordering things is a bit weird. I like your proposed UX flow just one thing I'd like to have - moving the selected starter to the top of the list.

So I've changed it as you proposed except modifiy list (not pushed yet):
- static typeahead input field with-out setting the starter url there
- move selected starter to top of list - I think that's needed as there are cases you won't see that you have selected a starter (e.g. new typeahead and selected starter on other pagination page or not in the results). Maybe we also need to add an info sign to the input field so it's clear that there is a starter selected (see details below `Indication of selected starter`)

**Indication of selected starter**
If the user selects a starter and then clears the typeahead input it's not clear that there is a starter selected.
OK, I have another solution to this. We won't hide the starters if typeahead is empty & toggling starter list to hidden is disabled if there is a selected starter.

**Loading of starters**
How should we handle a missing internet connection? I first thought that it would be good to have something like in the screenshot below. 
But a missing internet connection is a general problem for project creation and we should display that earlier (see console log below `unhandled promise` after clicking build with-out internet). I think disable build button & show a message like `Can't create a new project with-out internet connection` and/or a popover / heading on the modal with that info.
I think I'll remove the info from my working directory and we have to add this in a separate issue (issue not created yet). I think it's OK to fail silently for now. I'll create the issue in a minute.
![grafik](https://user-images.githubusercontent.com/3046542/51436888-7dee8800-1c95-11e9-9ed5-5e2a9aa874db.png)
![grafik](https://user-images.githubusercontent.com/3046542/51437116-bd1ed800-1c99-11e9-833b-dd1adfdf5a81.png)

AWolf81 commented:
@melanieseltzer Please have a look if this is OK and I have addressed everything. I think we can revisit and optimize this later.

I think the reordering of selected starter to the top is not perfect but OK for now. It would be better to keep the order and have a separate display of the selected starter. (Needs to be in a separate location because the typeahead could change the list and won't display the selected starter.)

Also list size is a bit small with just two items but I don't have an idea at the moment how we could easily optimize this. An idea would be to have icons for each build step and only display the active step or we could make the project icon selector collapsilble (we would probably get just one more item in the starter list with this as we could collapse to one icon).

melanieseltzer commented on src/components/CreateNewProjectWizard/Gatsby/SelectStarterList.js:
```diff
@@ -0,0 +1,177 @@
+                  <StarterItemHeading
+                    selected={selectedStarter === starter.repo}
+                    onClick={() =>
+                      this.handleUpdateStarter(
```
There's a lot of passing down/abstraction of the `updateFieldValue` function (from `CreateNewProjectWizard`) to ultimately end up here.

The chain right now seems to be:

1. `CreateNewProjectWizard` passes it to `MainPane` as `updateFieldValue` prop
2. `MainPane` passes it to `ProjectStarter` as `onSelect` prop
3. `ProjectStarter` passes it to `SelectStarterList` as `updateStarter` prop
4. `SelectStarterList` abstracts it into function `handleUpdateStarter`

This was a bit hard to follow when I was in `ProjectStarter` and wanted to see where `handleOnSelect` was getting its parameters set.

I wonder if there's a way to simplify this or at least keep the naming consistent when passing it around?

AWolf81 commented:
@melanieseltzer 
Oh, one point I've missed is that we have to update the `SummaryPane` for the starter as the user have to pick a starter from the list as entering the starter url into the input won't select the starter.

Or should we add that feature back in? E.g. if the user enteres a matching starter by name or url it will automatically pick the starter?
If the user enters a url that's not in the starters list it will be picked & used on build click - e.g. user wants to use an unlisted starter.
I'll add this and create a new PR for it.

melanieseltzer commented:
@AWolf81 Oh good point! I'll wait for the new PR because I've also identified an issue with search results that I'd like to iterate over.
